### PR TITLE
Token auth

### PR DESCRIPTION
**Pull request by weeellz**

### CODE REVIEW

File: HaroldCI.Api/Controllers/AccountController.cs [hunk 1]
```diff
@@ -25,50 +36,69 @@ public IActionResult Register()
...
+                var appUser = _userManager.Users.SingleOrDefault(r => r.UserName == model.Login);
+                return Json(new { Success = true, Token = await GenerateJwtToken(model.Login, appUser) });
             }
 
-            ModelState.AddModelError("", "Wrong login or password");
-            return View(model);
+            return Json(new { Success = false });
+        }
+
+        private async Task<object> GenerateJwtToken(string login, IdentityUser user)
```
Comment (fexolm) [inline HaroldCI.Api/Controllers/AccountController.cs]: There are no async operations here. Please change to sync call

File: HaroldCI.Api/Controllers/AccountController.cs [hunk 2]
```diff
@@ -25,50 +36,69 @@ public IActionResult Register()
...
+        {
+            var claims = new List<Claim>
+            {
+                new Claim(JwtRegisteredClaimNames.Sub, login),
+                new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString()),
+                new Claim(ClaimTypes.NameIdentifier, user.Id)
+            };
+
+            var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_configuration["JwtKey"]));
```
Comment (fexolm) [inline HaroldCI.Api/Controllers/AccountController.cs]: mb need to load jwtkey from file, current implementation is unsave

File: HaroldCI.Api/Startup.cs [hunk 3]
```diff
@@ -1,10 +1,15 @@
-﻿using HaroldCI.Models;
+﻿using System;
```
Comment (fexolm) [inline HaroldCI.Api/Startup.cs]: please sort usings (in whole project)

File: HaroldCI.Api/Controllers/AccountController.cs (hunk 2 above)
Comment (fexolm) [reply HaroldCI.Api/Controllers/AccountController.cs]: please replace JwtKey settings with JwtKeyFilename, and read key on startup

File: HaroldCI.Api/Startup.cs [hunk 4]
```diff
@@ -20,11 +25,32 @@ public Startup(IConfiguration configuration)
...
+                options.DefaultScheme = JwtBearerDefaults.AuthenticationScheme;
+                options.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
+            })
+                .AddJwtBearer(cfg =>
+            {
+                cfg.RequireHttpsMetadata = false;
+                cfg.SaveToken = true;
+                cfg.TokenValidationParameters = new TokenValidationParameters
+                {
+                    ValidIssuer = Configuration["JwtIssuer"],
```
Comment (fexolm) [inline HaroldCI.Api/Startup.cs]: Don't forget to change settings here

File: HaroldCI.Api/Views/Home/Index.cshtml [hunk 5]
```diff
@@ -1,6 +1,11 @@
-﻿@{
+﻿@model string
```
Comment (fexolm) [inline HaroldCI.Api/Views/Home/Index.cshtml]: remove changes in that file

File: HaroldCI.Api/Controllers/AccountController.cs [hunk 6]
```diff
@@ -25,50 +36,69 @@ public IActionResult Register()
         }
 
         [HttpPost]
-        [ValidateAntiForgeryToken]
-        public async Task<IActionResult> Register(UserView model)
+        public async Task<JsonResult> Register([FromBody] UserView model)
         {
-            if (!ModelState.IsValid) return View(model);
-            var user = new UserModel {UserName = model.Login};
+            var user = new UserModel
```
Comment (fexolm) [inline HaroldCI.Api/Controllers/AccountController.cs]: Please add model validation again

File: HaroldCI.Api/Controllers/AccountController.cs [hunk 7]
```diff
@@ -25,50 +36,69 @@ public IActionResult Register()
...
-            {
-                ModelState.AddModelError(String.Empty, error.Description);
-            }
-            return View(model);
+            return Json(new {Success=false});
         }
 
         [HttpGet]
         public IActionResult Login(string returnUrl = null)
         {
```
Comment (fexolm) [inline HaroldCI.Api/Controllers/AccountController.cs]: Remove unnecessary Login and Register views. They are not working anymore.

File: HaroldCI.Api/Controllers/AccountController.cs [hunk 8]
```diff
@@ -25,50 +36,69 @@ public IActionResult Register()
...
-            var result =
-                await _signInManager.PasswordSignInAsync(model.Login, model.Password, model.Remember, false);
-            if (result.Succeeded)
+            if (!ModelState.IsValid)
             {
-                if (!string.IsNullOrEmpty(model.ReturnUrl) && Url.IsLocalUrl(model.ReturnUrl))
-                {
-                    return Redirect(model.ReturnUrl);
-                }
+                return Json(new {Success = false, error=ModelState});
```
Comment (fexolm) [inline HaroldCI.Api/Controllers/AccountController.cs]: Need to add Response models, not just use anonymous structures

Comment (fexolm) [reply HaroldCI.Api/Controllers/AccountController.cs]: Also, please return error code on fail, not just 200

File: HaroldCI.Api/Controllers/AccountController.cs (hunk 1 above)
Comment (weeellz) [reply HaroldCI.Api/Controllers/AccountController.cs]: Done

File: HaroldCI.Api/Startup.cs (hunk 3 above)
Comment (weeellz) [reply HaroldCI.Api/Startup.cs]: Done

File: HaroldCI.Api/Views/Home/Index.cshtml (hunk 5 above)
Comment (weeellz) [reply HaroldCI.Api/Views/Home/Index.cshtml]: Done

File: HaroldCI.Api/Controllers/AccountController.cs (hunk 6 above)
Comment (weeellz) [reply HaroldCI.Api/Controllers/AccountController.cs]: Done

File: HaroldCI.Api/Controllers/AccountController.cs (hunk 7 above)
Comment (weeellz) [reply HaroldCI.Api/Controllers/AccountController.cs]: Done

File: HaroldCI.Api/Controllers/AccountController.cs (hunk 8 above)
Comment (weeellz) [reply HaroldCI.Api/Controllers/AccountController.cs]: Done

File: HaroldCI.Api/Controllers/AccountController.cs (hunk 2 above)
Comment (weeellz) [reply HaroldCI.Api/Controllers/AccountController.cs]: Done

File: HaroldCI.Api/Startup.cs (hunk 4 above)
Comment (weeellz) [reply HaroldCI.Api/Startup.cs]: Done

File: HaroldCI.Api/AuthOptions.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace HaroldCI.Api
+{
+    public class AuthOptions
+    {
+        public const string JwtIssuer = "http://localhost:62225/"; // издатель токена
+        public const string JwtAudience = "http://localhost:62225/"; // потребитель токена
+        public const string JwtKeyFilePath = @"C:\Users\weeel\Desktop\jwt.txt";   // ключ для шифрации
+        public const int JwtExpireDays = 30;
+
```
Comment (fexolm) [inline HaroldCI.Api/AuthOptions.cs]: Remove hardcoded settings

File: HaroldCI.Api/AuthOptions.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        public static string GetKey()
+        {
+            string key = "";
+
+            using (StreamReader file = new StreamReader(JwtKeyFilePath))
+            {
+                var temp = file.ReadLine();
+                key = temp;
+            }
+            return key;
```
Comment (fexolm) [inline HaroldCI.Api/AuthOptions.cs]: Read key from file just once

File: HaroldCI.Api/appsettings.Development.json [hunk 11]
```diff
@@ -9,8 +9,5 @@
   },
   "ConnectionStrings": {
     "DefaultConnection": "Server=localhost;Port=5432;Database=HaroldCI.Db;UserId=postgres;Password=Ravil97.;"
-  },
-  "JwtKey": "SOME_RANDOM_KEY_DO_NOT_SHARE",
-  "JwtIssuer": "http://localhost:62225/",
```
Comment (fexolm) [inline HaroldCI.Api/appsettings.Development.json]: Add settings again, but use JwtKeyFilePass, not JwtKey

**Comment by weeellz** [comment]
не знаю, как правильно написать Startup.cs

**Comment by weeellz** [comment]
seems working >_>

File: HaroldCI.Views/HaroldCI.Views.csproj [hunk 12]
```diff
@@ -4,4 +4,14 @@
     <TargetFramework>netcoreapp2.0</TargetFramework>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Newtonsoft.Json" Version="11.0.1" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Reference Include="Newtonsoft.Json">
+      <HintPath>..\..\..\..\..\..\..\program files\dotnet\sdk\NuGetFallbackFolder\newtonsoft.json\9.0.1\lib\netstandard1.0\Newtonsoft.Json.dll</HintPath>
```
Comment (fexolm) [inline HaroldCI.Views/HaroldCI.Views.csproj]: Something went wrong

Comment (weeellz) [reply HaroldCI.Views/HaroldCI.Views.csproj]: Done

**Pull request merged**